datafusion/physical-expr/src/aggregate/groups_accumulator.rs

```diff
@@ -0,0 +1,100 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Vectorized [`GroupsAccumulator`]
+
+use arrow_array::{ArrayRef, BooleanArray};
+use datafusion_common::Result;
+
+/// An implementation of GroupAccumulator is for a single aggregate
```

Comment:
Here is the new `GroupsAccumulator` trait that all accumulators would have to implement.

I would also plan to create a struct that implements this trait for aggregates based on `Accumulator` s

```rust
struct GroupsAdapter {
  groups: Vec<Box<dyn Accumulator>>
}

impl GroupsAccumulator for GroupsAdapter {
...
}
```

So in that way we can start with simpler (but slower) `Accumulator` implementations for aggregates, and provide a fast GroupsAccumulator for the aggregates / types that need the specialization